### PR TITLE
Forces gases to obey certain laws of nature

### LINKED ISSUE #33532: Division by zero by plasmic fusion
Issue reported from Round ID: 80176 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% LAG FREE])
image

The following pull requests are currently test merged:
#33460: 'Re-Support old hotkey mode behaviour' by ninjanomnom at commit d0e455
#33470: 'Refactors reagent container types' by ACCount12 at commit 9de85e
#33480: 'Deadmin tweaks: Admins without +AUTOLOGIN start deadmined. AUTOLOGIN defaults to on.' by MrStonedOne at commit 2ac9d8
#33465: 'Replaces a bunch of loc assignments with forcemoves and moves to nullspace' by vuonojenmustaturska at commit b82ed5

### PR DESCRIPTION
**Pull request by DaedalusGame**
:cl:
fix: Fixes Noblium Formation being multiplicative, so having 500 moles of tritium and 500 moles of nitrogen no longer produces 2500 moles of noblium (it should correctly produce 10 moles)
fix: Fixes most reactions deleting more gas than exists and making gas out of nowhere
fix: Fixes Stim Formation invoking a byond bug and not using its intended polynomial formula
fix: Fixes Cryo Cells not having garbage_collect and clamping
fix: Fixes Rad Collectors not having clamping
fix: Fixes Division by Zero when Fusion has no impurities.
fix: Removes a redundant line in lungs
fix: Fixes rad_act signal not firing from turfs, living mobs, rad collectors and geiger counters
/:cl:

I need someone to look over my math again because I'm not 100% sure this is right.
Things that should probably be done before merging this:
- ~~Fire, but it's complicated~~
- ~~Fusion, but it's disabled (Can we get a rework already damnit)~~
- Add clamping to everything that removes gases because obviously we can't trust garbage_collect()

Did you know that TRITIUM_BURN_TRIT_FACTOR is a constant factor for how insane tritium is?
Closes #33532

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 1]
```diff
@@ -207,13 +208,15 @@
 	var/old_heat_capacity = air.heat_capacity()
 	var/catalyst_efficency = max(min(cached_gases[/datum/gas/plasma][MOLES]/cached_gases[/datum/gas/tritium][MOLES],MAX_CATALYST_EFFICENCY)-(temperature/FUSION_HEAT_DROPOFF),1)
 	var/reaction_energy = THERMAL_ENERGY(air)
-	var/moles_impurities = air.total_moles()-(cached_gases[/datum/gas/plasma][MOLES]+cached_gases[/datum/gas/tritium][MOLES])
+	var/moles_impurities = max(air.total_moles()-(cached_gases[/datum/gas/plasma][MOLES]+cached_gases[/datum/gas/tritium][MOLES]),1)
 
 	var/plasma_fused = (PLASMA_FUSED_COEFFICENT*catalyst_efficency)*(temperature/PLASMA_BINDING_ENERGY)/10
 	var/tritium_catalyzed = (CATALYST_COEFFICENT*catalyst_efficency)*(temperature/PLASMA_BINDING_ENERGY)/40
+	plasma_fused = min(plasma_fused,cached_gases[/datum/gas/plasma][MOLES])
+	tritium_catalyzed = min(tritium_catalyzed,cached_gases[/datum/gas/tritium][MOLES])
```
Comment (vuonojenmustaturska) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: please combine these two assignments with the var initialization above

File: code/game/turfs/open.dm [hunk 2]
```diff
@@ -353,6 +353,7 @@
 
 /turf/open/rad_act(pulse_strength)
 	if (air.gases[/datum/gas/carbon_dioxide] && air.gases[/datum/gas/oxygen])
+		pulse_strength = min(pulse_strength,air.gases[/datum/gas/carbon_dioxide][MOLES]*1000,air.gases[/datum/gas/oxygen][MOLES]*2000)
```
Comment (duncathan) [inline code/game/turfs/open.dm]: could you leave a comment on these and similar limits elsewhere in the PR indicating that this is ensuring that matter is conserved appropriately?

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 3]
```diff
@@ -252,6 +255,7 @@
 
 	var/old_heat_capacity = air.heat_capacity()
 	var/heat_efficency = temperature/(FIRE_MINIMUM_TEMPERATURE_TO_EXIST*100)
+	heat_efficency = min(heat_efficency,cached_gases[/datum/gas/oxygen][MOLES],cached_gases[/datum/gas/nitrogen][MOLES])
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: as naksu said above

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 4]
```diff
@@ -284,6 +288,7 @@
 
 	var/old_heat_capacity = air.heat_capacity()
 	var/reaction_efficency = 1/((pressure/(0.1*ONE_ATMOSPHERE))*(max(cached_gases[/datum/gas/plasma][MOLES]/cached_gases[/datum/gas/tritium][MOLES],1)))
+	reaction_efficency = min(reaction_efficency,cached_gases[/datum/gas/tritium][MOLES],cached_gases[/datum/gas/plasma][MOLES]/2)
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: ^

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 5]
```diff
@@ -316,7 +321,8 @@
 	var/old_heat_capacity = air.heat_capacity()
 	var/heat_scale = air.temperature/STIMULUM_HEAT_SCALE
 	var/stim_energy_change
-	stim_energy_change =heat_scale + (STIMULUM_FIRST_RISE(heat_scale**2)) - (STIMULUM_FIRST_DROP(heat_scale**3)) + (STIMULUM_SECOND_RISE(heat_scale**4)) - (STIMULUM_ABSOLUTE_DROP(heat_scale**5))
+	heat_scale = min(heat_scale,cached_gases[/datum/gas/tritium][MOLES],cached_gases[/datum/gas/plasma][MOLES],cached_gases[/datum/gas/nitryl][MOLES])
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: ^

File: code/modules/atmospherics/gasmixtures/reactions.dm [hunk 6]
```diff
@@ -345,7 +351,8 @@
 	var/list/cached_gases = air.gases
 	air.assert_gases(/datum/gas/hypernoblium,/datum/gas/bz)
 	var/old_heat_capacity = air.heat_capacity()
-	var/nob_formed = (cached_gases[/datum/gas/nitrogen][MOLES]*cached_gases[/datum/gas/tritium][MOLES])/100
+	var/nob_formed = (cached_gases[/datum/gas/nitrogen][MOLES]+cached_gases[/datum/gas/tritium][MOLES])/100
+	nob_formed = min(nob_formed,cached_gases[/datum/gas/tritium][MOLES]/10,cached_gases[/datum/gas/nitrogen][MOLES]/20)
```
Comment (duncathan) [inline code/modules/atmospherics/gasmixtures/reactions.dm]: ^

**Review by duncathan — CHANGES_REQUESTED** [review]
doing these in one line is ugly but unfortunately necessary for the minor performance enhancements that come with it

**Pull request merged**